design/clean_artifacts_in_csi_flow.md

```diff
@@ -0,0 +1,422 @@
+
+	if vscFromBackup.Status != nil && vscFromBackup.Status.SnapshotHandle != nil {
+		snapCont.Spec.Source.VolumeHandle = nil
+		snapCont.Spec.Source.SnapshotHandle = vscFromBackup.Status.SnapshotHandle
+	} else {
+		p.log.Errorf("fail to get snapshot handle from VSC %s status", snapCont.Name)
+		return nil, errors.Errorf("fail to get snapshot handle from VSC %s status", snapCont.Name)
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+	if csi.IsVolumeSnapshotContentHasDeleteSecret(&snapCont) {
+		additionalItems = append(additionalItems,
+			velero.ResourceIdentifier{
+				GroupResource: schema.GroupResource{Group: "", Resource: "secrets"},
+				Name:          snapCont.Annotations[velerov1api.PrefixedSecretNameAnnotation],
+				Namespace:     snapCont.Annotations[velerov1api.PrefixedSecretNamespaceAnnotation],
+			},
+		)
+	}
+
+	vscMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&snapCont)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	p.log.Infof("Returning from VolumeSnapshotContentRestoreItemAction with %d additionalItems",
+		len(additionalItems))
+	return &velero.RestoreItemActionExecuteOutput{
+		UpdatedItem:     &unstructured.Unstructured{Object: vscMap},
+		AdditionalItems: additionalItems,
+	}, nil
+}
+```
+
+### Backup Sync
+csi-volumesnapshotclasses.json, csi-volumesnapshotcontents.json, and csi-volumesnapshots.json are CSI-related metadata files in the BSL for each backup.
+
+csi-volumesnapshotcontents.json and csi-volumesnapshots.json are not needed anymore, but csi-volumesnapshotclasses.json is still needed.
+
+One concrete scenario is that a backup is created in cluster-A, then the backup is synced to cluster-B, and the backup is deleted in the cluster-B. In this case, we don't have a chance to create the VolumeSnapshot and VolumeSnapshotContent needed VolumeSnapshotClass.
+
+As a result, the VolumeSnapshotClass should still be synced in the backup sync process.
+
+### Backup Deletion
+Two factors are worthy for consideration for the backup deletion change:
+* Because the VolumeSnapshotContents generated by the backup are not synced anymore, and the VolumeSnapshotContents generated during the backup will not be kept too. The backup deletion needs to generate a VolumeSnapshotContent, then deletes it to make sure the snapshots in the storage provider are clean too.
+* The VolumeSnapshots generated by the backup are already deleted in the backup process, we don't need a DeleteItemAction for the VolumeSnapshot anymore. As a result, the `velero.io/csi-volumesnapshot-delete` plugin is unneeded.
+
+For the VolumeSnapshotContent DeleteItemAction, we need to generate a VolumeSnapshotContent. Because we only care about the snapshot deletion, we don't need to create a VolumeSnapshot associated with the VolumeSnapshotContent.
+
+Create a static VolumeSnapshotContent, then point it to a pseudo VolumeSnapshot, and reference to the snapshot handle should be enough.
```

Comment:
Do we still need the pseudo VolumeSnapshot? What about delete the VolumeSnapshotContent directly when its `deletionPolicy` is `Delete`

Reply:
This was a misunderstanding. The design didn't mean to create a VS too. There is no need to create the pseudo VS.